lib/chef/provider/launchd.rb

```diff
@@ -0,0 +1,198 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/provider'
+require 'chef/resource'
+require 'chef/resource/file'
+require 'chef/resource/cookbook_file'
+require 'chef/resource/macosx_service'
+
+class Chef
+  class Provider
+    class Launchd < Chef::Provider
+
+      def load_current_resource
+        @current_resource = Chef::Resource::Launchd.new(@new_resource.name)
+        @name = @new_resource.name
+        @label = @new_resource.label ? @new_resource.label : @name
+        @backup = @new_resource.backup
+        @cookbook = @new_resource.cookbook
+        @group = @new_resource.group
+        @mode = @new_resource.mode
+        @owner = @new_resource.owner
+        @path = @new_resource.path
+        @source = @new_resource.source
+        @content = content?
+      end
+
+      def action_create
+        manage_plist(:create)
+      end
+
+      def action_create_if_missing
+        manage_plist(:create_if_missing)
+      end
+
+      def action_delete
+        # If you delete a service you want to make sure its not loaded or
+        # the service will be in memory and you wont be able to stop it.
+        if ::File.exists?(@path)
+          manage_service(:disable)
+        end
+        manage_plist(:delete)
+      end
+
+      def action_enable
+        if manage_plist(:create)
+          manage_service(:restart)
+        else
+          manage_service(:enable)
+        end
+      end
+
+      def action_disable
+        manage_service(:disable)
+      end
+
+      def manage_plist(action)
+        if @source
+          res = cookbook_file_resource
+        else
+          res = file_resource
+        end
+        res.run_action(action)
+        new_resource.updated_by_last_action(true) if res.updated?
+        res.updated
+      end
+
+      def manage_service(action)
+        res = service_resource
+        res.run_action(action)
+        new_resource.updated_by_last_action(true) if res.updated?
+      end
+
+      def service_resource
+        res = Chef::Resource::MacosxService.new(@label, run_context)
+        res.name(@label)
+        res.service_name(@label)
+        res.plist(@path)
+        res.session_type(@session_type) if @session_type
+
+        res
+      end
+
+      def file_resource
+        res = Chef::Resource::File.new(@path, run_context)
+        res.name(@path)
+        res.backup(@backup) if @backup
+        res.content(@content) if @content
+        res.group(@group) if @group
+        res.mode(@mode) if @mode
+        res.owner(@owner) if @owner
+
+        res
+      end
+
+      def cookbook_file_resource
+        res = Chef::Resource::CookbookFile.new(@path, run_context)
+        res.cookbook_name = @cookbook if @cookbook
+        res.name(@path)
+        res.backup(@backup) if @backup
+        res.group(@group) if @group
+        res.mode(@mode) if @mode
+        res.owner(@owner) if @owner
+        res.source(@source) if @source
```

Comment:
all this kind of stuff should be using just `source` which should be an accessor delegated to `new_resource.source`.